sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java

```diff
@@ -0,0 +1,222 @@
+import java.util.List;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoder;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.test.TestSparkSession;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+public class JavaBeanWithArraySuite {
+
+    private static final List<Record> RECORDS = new ArrayList<>();
+
+    static {
+        RECORDS.add(new Record(1,
+                Arrays.asList(new Interval(111, 211), new Interval(121, 221)),
+                Arrays.asList(11, 21, 31, 41)
+        ));
+        RECORDS.add(new Record(2,
+                Arrays.asList(new Interval(112, 212), new Interval(122, 222)),
+                Arrays.asList(12, 22, 32, 42)
+        ));
+        RECORDS.add(new Record(3,
+                Arrays.asList(new Interval(113, 213), new Interval(123, 223)),
+                Arrays.asList(13, 23, 33, 43)
+        ));
+    }
+
+    private TestSparkSession spark;
+
+    @Before
+    public void setUp() {
+        spark = new TestSparkSession();
+    }
+
+    @After
+    public void tearDown() {
+        spark.stop();
+        spark = null;
+    }
+
+    @Test
+    public void testBeanWithArrayFieldsDeserialization() {
+
+        StructType schema = createSchema();
+        Encoder<Record> encoder = Encoders.bean(Record.class);
+
+        Dataset<Record> dataset = spark
+                .read()
+                .format("json")
+                .schema(schema)
+                .load("src/test/resources/test-data/with-array-fields")
+                .as(encoder);
+
+        List<Record> records = dataset.collectAsList();
+
+        Assert.assertTrue(Util.equals(records, RECORDS));
+    }
+
+    private static StructType createSchema() {
+        StructField[] intervalFields = {
+                new StructField("startTime", DataTypes.LongType, true, Metadata.empty()),
+                new StructField("endTime", DataTypes.LongType, true, Metadata.empty())
+        };
+        DataType intervalType = new StructType(intervalFields);
+
+        DataType intervalsType = new ArrayType(intervalType, true);
+
+        DataType valuesType = new ArrayType(DataTypes.IntegerType, true);
+
+        StructField[] fields = {
+                new StructField("id", DataTypes.IntegerType, true, Metadata.empty()),
+                new StructField("intervals", intervalsType, true, Metadata.empty()),
+                new StructField("values", valuesType, true, Metadata.empty())
+        };
+        return new StructType(fields);
+    }
+
+    public static class Record {
+
+        private int id;
+        private List<Interval> intervals;
+        private List<Integer> values;
```

Comment:
Will this list of int affect the test? If no, maybe we can get rid of it to simplify the test.

Reply:
The intention was to test a non struct case too. But I think, it's really not critical and we can get rid of it.